formats/cbor/commonMain/src/kotlinx/serialization/cbor/CborElement.kt

```diff
@@ -0,0 +1,300 @@
+ *
+ * [CborElement.toString] properly prints CBOR tree as a human-readable representation.
+ * Whole hierarchy is serializable, but only when used with [Cbor] as [CborElement] is purely CBOR-specific structure
+ * which has meaningful schemaless semantics only for CBOR.
+ *
+ * The whole hierarchy is [serializable][Serializable] only by [Cbor] format.
+ */
+@Serializable(with = CborElementSerializer::class)
+@ExperimentalSerializationApi
+public sealed class CborElement(
+    /**
+     * CBOR tags associated with this element.
+     * Tags are optional semantic tagging of other major types (major type 6).
+     * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+     */
+    @OptIn(ExperimentalUnsignedTypes::class)
+    tags: ULongArray = ulongArrayOf()
+
+) {
+    /**
+     * CBOR tags associated with this element.
+     * Tags are optional semantic tagging of other major types (major type 6).
+     * See [RFC 8949 3.4. Tagging of Items](https://datatracker.ietf.org/doc/html/rfc8949#name-tagging-of-items).
+     */
+    @OptIn(ExperimentalUnsignedTypes::class)
+    public var tags: ULongArray = tags
+        internal set //need this to collect
+
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is CborElement) return false
+
+        if (!tags.contentEquals(other.tags)) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        return tags.contentHashCode()
+    }
+
+}
+
+/**
+ * Class representing CBOR primitive value.
+ * CBOR primitives include numbers, strings, booleans, byte arrays and special null value [CborNull].
+ */
+@Serializable(with = CborPrimitiveSerializer::class)
+@ExperimentalSerializationApi
+public sealed class CborPrimitive<T : Any>(
+    public val value: T,
+    tags: ULongArray = ulongArrayOf()
+) : CborElement(tags) {
+    override fun equals(other: Any?): Boolean {
+        if (this === other) return true
+        if (other !is CborPrimitive<*>) return false
+        if (!super.equals(other)) return false
+
+        if (value != other.value) return false
+
+        return true
+    }
+
+    override fun hashCode(): Int {
+        var result = super.hashCode()
+        result = 31 * result + value.hashCode()
+        return result
+    }
+
+    override fun toString(): String {
+        return "${this::class.simpleName}(" +
+            "tags=${tags.joinToString(prefix = "[", postfix = "]")}, " +
+            "value=$value" +
+            ")"
+    }
+}
+
+/**
+ * Class representing either:
+ * * signed CBOR integer (major type 1 encompassing `-2^64..-1`)
+ * * unsigned CBOR integer (major type 0 encompassing `0..2^64-1`)
+ *
+ * depending on the value of [sign]. Note that [absoluteValue] **must** be `0` when sign is set to [Sign.ZERO]
+ */
+@Serializable(with = CborIntSerializer::class)
+@ExperimentalSerializationApi
+public class CborInt(
```

Comment:
What do you think about `CborInteger` instead? It's closer to major type descriptions from the spec (`an unsigned integer`, `a negative integer`) and it is farther from Kotlin's `Int`.

Reply:
I'm fine with both. Completely up to you guys

Reply:
Let's make it `CborInteger` then.